### ISSUE TITLE
How to unit test workflow's after_commit hook's correctly?

### ISSUE THREAD

**Issue by nmacuk** [zilverline/sequent#333]
Hello!
I have stumbled upon an issue when testing workflows

In particular i’m talking about `after_commit` hook’s.

In the `FakeTransactionProvider` that is used inside tests, everything that is queued as `@after_commit_blocks` is not executed anywhere. The only place I saw when the method responsible for actually executing the blocks is `transactional` method inside `FakeTransactionProvider`

```ruby
class FakeTransactionProvider
  def initialize
    @after_commit_blocks = []
  end

  def transactional
    yield
    @after_commit_blocks.each(&:call)
  end

  def after_commit(&block)
    @after_commit_blocks << block
  end
end
```

But from looking into the source code, the `transactional` method is only called inside `lib/sequent/core/command_service.rb` when processing commands.

This makes sense if the workflow is "triggered" by command handler, which wraps the entire operation within transaction. But this implies the problem with unit testing workflows. The `after_commit` block is never called in such scenarios. 

Let’s say I have a Workflow like this:

```ruby
class NotifyNewUserCreated < Sequent::Workflow
  on UserCreated do |event|
    after_commit do
      Notifier::UserCreated.new(user_uuid: event.user_uuid, email: event.user_email).call
    end
  end
end
```

If I was writing a spec for the workflow I would do something like this:

```ruby
require "rails_helper"

describe NotifyNewUserCreated, type: :workflow do
  let(:workflow) { described_class.new }

  before do
    Sequent.configuration.event_handlers = [described_class.new]
  end

let(:user_uuid) { Sequent.new_uuid }
let(:user_email) { FFaker::Internet.email }
let(:user_created) { double(Notifier::UserCreated) }

  describe "on UserCreated" do
    it "notifies about new user" do
      when_event(UserCreated.new(...))
       
      allow(Notifier::UserCreated).to receive(:new).with(user_uuid: user_uuid, email: user_email).and_return(user_created)
      expect(user_created).to receive(:call)
    end
  end
end
```

This does not work because the `after_commit` block is not executed.

What I did is that I created a helper like this:

```ruby
module WorkflowHelper
  def then_after_commit(&block)
    yield
    Sequent.configuration.transaction_provider.instance_variable_get(:@after_commit_blocks).each(&:call)
  end
end
```

and then in spec I can do this:

```ruby
describe "on UserCreated" do
  it "notifies about new user" do
    when_event(UserCreated.new(...))
    then_after_commit do
      allow(Notifier::UserCreated).to receive(:new).with(user_uuid: user_uuid, email: user_email).and_return(user_created)
      expect(user_created).to receive(:call)
    end
  end
end
```

But this helper is too tightly coupled to the underlying implementation of `transaction_provider` and could easily break with possible changes to Sequent. 

In summary, how should one handle testing workflows with `after_commit` hooks, when the `FakeTransactionProvider` used for tests doesn't execute any `after_commit` hooks?

**Comment by lvonk** [comment]
Hi @nmacuk,

That indeed would work, but I agree that is too tightly coupled. Since the `after_commit` hooks only run when a "commit" is done, I guess the best thing todo is to wrap the test in a transactions and call the `FakeTransactionProvider` manually:

```
describe "on UserCreated" do
  around do |example|
    # where transaction_provider = FakeTransactionProvider
    Sequent.configuration.transaction_provider.transactional do
      example.run
    end
  end

  it "notifies about new user" do
      allow(Notifier::UserCreated).to receive(:new).with(user_uuid: user_uuid, email: user_email).and_return(user_created)
      expect(user_created).to receive(:call)

    when_event(UserCreated.new(...))
  end
end
```

Would this work for you?

**Comment by nmacuk** [comment]
Thanks for responding so quickly.
Yes, that indeed has worked for us.

I saw that inside the `transactional` block there are commands processed, but did not realize that I could swap it and execute events as well.

Anyway, thank you very much and probably see you around here from time to time 😄

**Comment by lvonk** [comment]
Great, happy to help.
